README.md
R utility functions for Arpa ER air quality data


NEWS
Changes in arpautils 0.9.6 [2017-01-30]
=====================================
 * dbqa.get.idcfgsens: risolto piccolo bug che si poteva manifestare al cambio di sensore
 
 
Changes in arpautils 0.9.5 [2016-05-18]
=====================================
 * calculate.ozone_annual_report: corretto bug in calcolo efficienza giornaliera
 
 
Changes in arpautils 0.9.4 [2016-04-22]
=====================================
 * write.daily_report: gestisce meglio i mancanti; sovrascrive; bug fixed
 
 
Changes in arpautils 0.9.3 [2016-04-08]
=====================================
 * corretto annual report (mancava un arrotondamento)
 * aggiunte funzioni prepare.daily_report, calculate.daily_report, write.daily_report
 
 
Changes in arpautils 0.9.2 [2016-04-04]
=====================================
 * aggiunta media annua del max giorn.media 8h per CO (id.elab=821)
 
 
Changes in arpautils 0.9.1 [2016-02-22]
=====================================
 * corretto errore in dbqa.descr.elab
 
 
 Changes in arpautils 0.9.0 [2016-01-19]
=====================================
 * modifiche a write.annual_report: ora scrive il numero di dati usati e gestisce casi <LOD
 * modifiche a write.ozone_annual_report: ora scrive il numero di dati usati
 * compatibilità con sintassi S3: mean.window rinominata in mean_window, round.awayfromzero rinominata in round_awayfromzero
 * rivista la gestione di Imports in DESCRIPTION e NAMESPACE


Changes in arpautils 0.8.17 [2016-01-14]
=====================================
 * nuove funzioni: dbqa.lod, dbqa.get.idparam, dbqa.get.elab, dbqa.descr.elab
 * dbqa.config ora permette di leggere le credenziali da un file e di esportarle come variabili globali
 * dbqa.connect adeguato di conseguenza
 * dbqa.get.datastaz gestisce il limit of detection
 

Changes in arpautils 0.8.16 [2015-11-03]
=====================================
 * calculate.ozone_annual_report restituisce anche annual.mean, aot40.veget.NhValid, aot40.forest.NhValid
 * aot restituisce anche NhValid


Changes in arpautils 0.8.15 [2015-10-12]
=====================================
 * detect.event restituisce anche index.max, indice del [primo] massimo nell'evento


Changes in arpautils 0.8.14 [2015-05-18]
=====================================
 * dbqa.list.active.staz include laboratori mobili, se richiesto


Changes in arpautils 0.8.13 [2015-03-20]
=====================================
 * timezone da BST a Africa/Algiers


Changes in arpautils 0.8.12 [2015-01-28]
=====================================
 * corretto bug nell'AOT 


Changes in arpautils 0.8.11 [2015-01-19]
=====================================
 * corretto bug nella valutazione della validita' delle statistiche annuali ozono


Changes in arpautils 0.8.10 [2015-01-15]
=====================================
 * gestione selezione stazioni RRQA con dbqa.isrrqa()
 * estrazione tipo_emis e tipo_area


Changes in arpautils 0.8.9 [2015-01-09]
=====================================
 * debug negli arrotondamenti in calculate.ozone_annual_report e calculate.ozone_daily_report


Changes in arpautils 0.8.8 [2014-12-18]
=====================================
 * reintrodotto arrotondamento del max media mobile (non era un bug!)
 * aggiustamenti minimi nel manuale


Changes in arpautils 0.8.7 [2014-12-05]
=====================================
 * adeguato l'arrotondamento (0.5 diventa 1)
 * arrotonda l'efficienza annuale (come UE)
 * dbqa.get.datasens accede a AA_ARIA.T$01$CONFIG_SENSORI (piu' veloce)


Changes in arpautils 0.8.6 [2014-12-03]
=====================================
 * migliora scrittura FLG_ELAB
 * considera VAL_OFFSET in estrazione da DB
 * arrotonda in visualizzazione in dbqa.get.datasens
 * tolto arrotondamento della media mobile (bug)


Changes in arpautils 0.8.5 [2014-12-01]
=====================================
 * aggiunta dbqa.round() per gestire gli arrotondamenti come prescritto dal GdL
 * aggiunta write.annual_report()
 * aggiunta Ndays.in.year()
 * corretto bug nel calcolo dell'efficienza


Changes in arpautils 0.8.4 [2014-11-24]
=====================================
 * aggiunti a calculate.annual_report(): superamenti orari, orari consecutivi, media di mesi selezionati, calcolo dati attesi ed efficienza


Changes in arpautils 0.8.3 [2014-11-12]
=====================================
 * migliorata gestione delle flag di validazione in dbqa.get.datasens()
 * completata la gestione delle statistiche annuali ozono con write.ozone_annual_report()


Changes in arpautils 0.8.2 [2014-11-10]
=====================================
 * corretto bug nella scrittura della durata dell'evento in write.ozone_annual_report()


Changes in arpautils 0.8.1 [2014-10-31]
=====================================
 * gestione piu' flessibile delle flag di validazione in dbqa.get.datasens()


Changes in arpautils 0.8.0 [2014-10-27]
=====================================
 * rimosse dal pacchetto le credenziali di accesso al DB


Changes in arpautils 0.7.1 [2014-10-24]
=====================================
 * aggiunto Sys.setenv(TZ="BST") in dbqa.config()
 * standardizzata sintassi delle stringhe che definiscono i formati data/ora SQL
 * aggiunte le funzioni prepare.ozone_annual_report(), calculate.ozone_annual_report()


Changes in arpautils 0.6.1 [2014-10-10]
=====================================
 * modificata dbqa.insert per correggere errore di scrittura dei campi DATE nel DB
 * aggiunte le funzioni prepare.annual_report, calculate.annual_report


Changes in arpautils 0.5.3 [2014-07-21]
=====================================
 * prepare.ozone_daily_report accede a tabella dati recenti
 * aggiornata documentazione


R/xts.functions.R
## converte i dati come escono dal DBQA in un oggetto xts
dbqa.data2xts <- function(data,
                          Date="TS_INIZIO_RIL",
                          Value="VALORE",
                          TZ="BST") {
  out <- xts(data[[Value]], as.POSIXct(data[[Date]], tz=TZ), tzone=TZ) ## sembra sia l'unico modo per codificare UTC+1 senza DST
  return(out)
}


## regolarizza una serie temporale xts,
## a passi orari o giornalieri
xts.regolarize <- function(tstep, x,
                           f.time=(i<-index(x))[1],
                           l.time=i[length(i)],
                           TZ="BST") {
  ## controlla lo step
  if(!(tstep%in%c("d","H"))) stop(paste("Invalid timestep value, only `d` or `H` allowed"))
  
  ## costruisce la sequenza temporale regolare
  timestring <- paste(format(f.time,"%Y-%m-%d %H:%M"),
                      format(l.time,"%Y-%m-%d %H:%M"),
                      tstep,sep="/")
  Sys.setenv(TZ=TZ)
  s0 <- timeBasedSeq(timestring)
  
  ## costruisce un oggetto xts vuoto regolare
  x0 <- xts(rep(NA,length(s0)), s0, tzone=TZ)
  
  ## sincronizza l'oggetto xts con quello regolare vuoto
  xtmp <- merge.xts(x0,x,join="left", tzone=TZ)
  xreg <- xts(xtmp$x, index(xtmp), tzone=TZ)
  
  return(xreg)
}

## unisce molte serie temporali in una sola,
## regolarizzandole a passi orari o giornalieri
xts.blend <- function(tstep, TZ="BST", ...) {
  ## controlla lo step
  if(!(tstep%in%c("d","H"))) stop(paste("Invalid timestep value, only `d` or `H` allowed"))
  
  ## mette le serie temporali in un solo oggetto xts complessivo
  Xm <- merge.xts(..., tzone=TZ)
  
  ## costruisce la sequenza temporale regolare
  idx.time <- index(Xm)
  f.time <- idx.time[1]
  l.time <- idx.time[length(idx.time)]
  timestring <- paste(format(f.time,"%Y-%m-%d %H:%M"),
                      format(l.time,"%Y-%m-%d %H:%M"),
                      tstep,sep="/")
  s0 <- timeBasedSeq(timestring)
  
  ## costruisce un oggetto xts vuoto regolare
  x0 <- xts(rep(NA,length(s0)), s0, tzone=TZ)
  
  ## sincronizza l'oggetto xts complessivo con quello regolare vuoto
  Xm <- merge.xts(x0,Xm,join="left", tzone=TZ)
  
  ## tiene un solo valore per ogni passo temporale
  ## (se ne trova 2 o più restituisce mancante)
  alone <- function(x) {
    ifelse(test = sum(!is.na(x))==1,
           yes  = max(x,na.rm=T),
           no   = NA)
  }
  dm <- apply(Xm,MARGIN=1,FUN=alone)
  #dm[is.infinite(dm)] <- NA
  #dm[is.nan(dm)] <- NA
  xm <- xts(dm, s0)
  
  return(xm)
}


R/time.functions.R
Hour   <- function(x,tz="BST") as.numeric(format(as.POSIXct(x, tz=tz), format="%H"))

Month  <- function(x,tz="BST") as.numeric(format(as.POSIXct(x, tz=tz), format="%m"))

Year   <- function(x,tz="BST") as.numeric(format(as.POSIXct(x, tz=tz), format="%Y"))

Ymd    <- function(x,tz="BST") as.numeric(format(as.POSIXct(x, tz=tz), format="%Y%m%d"))

Ym     <- function(x,tz="BST") as.numeric(format(as.POSIXct(x, tz=tz), format="%Y%m"))

YQ     <- function(x,tz="BST") paste(Year(x,tz=tz),quarters(x),sep="")

Ndays  <- function(x,tz="BST") length(unique(Ymd(x,tz=tz)))

Nmonths<- function(x,tz="BST") length(unique(Ym(x,tz=tz)))

Ndays.in.year <- function(year,tz="BST") {
  nd <- length(seq(as.POSIXct(paste(year,"-01-01",sep=""),tz=tz),
                   as.POSIXct(paste(year,"-12-31",sep=""),tz=tz),
                   by="1 days"))
  return(nd)
}

squeeze<- function(x) {
  if(length(x)==0) {
    out <- "-"
  } else {
    x <- sort(x,decreasing=F)
    xb <- c(NA,x)[1:length(x)]
    xa <- c(x,NA)[2:(1+length(x))]
    ext <- (xa-x)>1 | (x-xb)>1
    sel <- which(is.na(ext)|ext)
    x.sel <- x[sel]
    sela <- c(sel,NA)[2:(1+length(sel))]
    sym <- c("-",",")[as.numeric(sela-sel==1)+1]
    sym[is.na(sym)] <- ""    
    out <- paste(rbind(x.sel,sym),collapse="")
  }
  return(out)
}


R/aqstat.functions.R
# contiene funzioni utili per statistiche di dati QA
# con valenza legale o meno

# stat.minvalid <- function(x,necess,fun=mean,...) {
#   xmean  <- c(NA,
#               fun(x,...))[1+as.integer(sum(!is.na(x)>=necess))]
#   return(xmean)
# }

stat.period <- function(x,period,necess,FUN=mean) {
  xmean  <- suppressWarnings(tapply(X=x, INDEX=period, FUN=FUN, na.rm=T))
  xvalid <- tapply(X=!is.na(x),
                   INDEX=period, FUN=sum, na.rm=T)
  xmean[xvalid < necess] <- NA
  return(xmean)
}

which.period <- function(x,period,necess,FUN=which.max) {
  xwhich  <- suppressWarnings(tapply(X=x, INDEX=period, FUN=FUN))
  xvalid <- tapply(X=!is.na(x),
                   INDEX=period, FUN=sum, na.rm=T)
  xwhich[xvalid < necess] <- NA
  return(xwhich)
}

stat.period2 <- function(x,period,nmax.missing,FUN=mean) {
  xmean  <- suppressWarnings(tapply(X=x, INDEX=period, FUN=FUN, na.rm=T))
  xinvalid <- tapply(X=is.na(x),
                   INDEX=period, FUN=sum, na.rm=T)
  xmean[xinvalid > nmax.missing] <- NA
  return(xmean)
}

shift <- function(x,k) {
  lx <- length(x)
  if(k>0) {
    out <- c(rep(NA,k), x)[1:lx]
  }
  if(k<0) {
    out <- c(x, rep(NA,-k))[(1-k):(lx-k)]
  }
  if(k==0) {
    out <- x
  }
  return(out)
}

stat.window <- function(x,window,necess,FUN=mean) {
  X <- NULL
  for(i in window[1]:window[2]) {
    xx <- shift(x,-i)
    X <- rbind(X,xx)
  }
  xmean <- suppressWarnings(apply(X, MARGIN=2, FUN=FUN, na.rm=T))
  xvalid <- apply(X=!(is.na(X)), MARGIN=2,
                  FUN=sum, na.rm=T)
  xmean[xvalid < necess] <- NA
  return(xmean)
}

mean.window <- function(x,k,necess) {
  xmean <- runmean(x,k=k,alg="C",endrule="NA",align="right")
  xvalid <- runmean(!is.na(x),k=k,alg="C",endrule="NA",align="right")*k
  xmean[xvalid < necess] <- NA
  return(xmean)
}

exc.period <- function(x,period,necess,threshold){
  xx <- x > threshold
  out <- stat.period(xx,period=period,
                     necess=necess,FUN=sum)
  return(out)
}

detect.event <- function(x,threshold) {
  over <- (x>threshold)
  over[is.na(over)] <- F
  event.start <- over & !c(F,over)[1:length(over)]
  index.event.start <- which(event.start)
  id.event <- cumsum(event.start)*over
  n.events <- max(id.event)
  if(n.events>0) {
    event.duration  <- NULL
    event.max       <- NULL
    index.event.max <- NULL
    for(i in 1:n.events) {
      event.duration  <- c(event.duration, sum(i==id.event))
      event.max       <- c(event.max,      max(x[i==id.event]))
      index.event.max <- c(index.event.max,index.event.start[i] + which.max(x[i==id.event]) - 1)
    }
    out <- data.frame(index    = index.event.start,
                      duration = event.duration,
                      max      = event.max,
                      index.max= index.event.max)
  } else {
    out <- NULL
  }
  return(out)
}

aot <- function(x, hr, threshold=80, estimate=T, hr.min=8, hr.max=19) {
  if(length(x)!=length(hr)) stop("x and hr must have the same length!")
  in.hr <- hr>=hr.min & hr<=hr.max
  valid <- !is.na(x)
  in.hr.valid <- in.hr&valid
  delta.positive <- pmax(x[in.hr]-threshold,0)
  Aot <- sum(delta.positive, na.rm=T)
  if(estimate) Aot <- Aot*sum(in.hr)/sum(in.hr.valid)
  NhValid <- sum(in.hr.valid)
  PercValid <- NhValid/sum(in.hr)*100
  out <- list(Aot=Aot, PercValid=PercValid, NhValid=NhValid)
  return(out)
}



R/dbqa.functions.R
## Funzioni di accesso e interrogazione al DB QA

## credenziali accesso DB
dbqa.config <- function(db_usr=NULL, 
                        db_pwd=NULL, 
                        db_name=NULL, 
                        db_tz="Africa/Algiers",
                        config.file="~/util/R/dbqa_keys.R",
                        global=TRUE) {
  if(is.null(db_usr)) {
    if(!file.exists(config.file)) stop(paste("Cannot find file",config.file))
    cat(paste("Reading dbqa configuration from file",config.file),sep = "\n")
    source(config.file,local = TRUE)
  }
  if(global) {
    cat("Setting dbqa configuration globally",sep = "\n")
    db_usr  <<- db_usr
    db_pwd  <<- db_pwd
    db_name <<- db_name
    db_tz   <<- db_tz
    Sys.setenv(TZ=db_tz)
  }
  cfg <- list(db_usr=db_usr, db_pwd=db_pwd, db_name=db_name, db_tz=db_tz)
  return(cfg)
}

## funzione per connettersi al DB di qualita' dell'aria Arpa ER
## (Africa/Algiers = ora locale italiana senza DST)
dbqa.connect <- function(...) { 
  ## Carichiamo la libreria Oracle
  ## Carichiamo il driver
  drv <- dbDriver("Oracle")
  ## credenziali accesso DB
  cfg <- dbqa.config(...)
  ## Creiamo la connessione al DB
  con <- dbConnect(drv,
                   username = cfg$db_usr,
                   password = cfg$db_pwd,
                   dbname   = cfg$db_name)
  return(con)
}


## scrive in una stringa una semplice query della forma
## "SELECT what FROM tab [WHERE crit]"
simple.query <- function(tab,what,crit) {
  out <- paste("SELECT",paste(what,collapse=","),
               "FROM",paste(tab,collapse=","),
               ("WHERE")[!is.null(crit)],paste(crit,collapse=" AND "))
  return(out)
}


## dato un ID parametro e sensore,
## scarica i dati disponibili cosi' come sono
dbqa.get.datasens <- function(con,
                              ts.range,
                              id.cfgsens,
                              id.param,
                              flg=1,
                              flg.excl=NULL,
                              flg.null=FALSE,
                              verbose=FALSE,
                              table="storico",
                              ...) {
  
  ## costruisce query
  if(table=="storico") {
    tab <- c("AA_ARIA.T$01$DATI_STORICO d",
             "AA_ARIA.T$01$CONFIG_SENSORI cs")    
  } else {
    tab <- c("AA_ARIA.T$01$DATI d",
             "AA_ARIA.T$01$CONFIG_SENSORI cs")    
    
  }
  what <- c("TO_CHAR(d.TS_INIZIO_RIL,'YYYY-MM-DD HH24:MI') as TS_INIZIO_RIL",          ## OCCHIO CON LA DATA, CONTROLLARE
            "d.VALORE * cs.COEFF_CONV + cs.VAL_OFFSET as VALORE",
            "d.FLG_A as FLG_A")
  crit <- c(paste("d.ID_PARAMETRO=",id.param,sep=""),
            paste("d.ID_CONFIG_SENSORE=",id.cfgsens,sep=""),
            "d.ID_CONFIG_SENSORE=cs.ID_CONFIG_SENSORE",
            paste("d.TS_INIZIO_RIL between to_date('",
                  format(as.POSIXct(ts.range[1]),"%Y-%m-%d %H:%M"),
                  "','YYYY-MM-DD HH24:MI') and to_date('",
                  format(as.POSIXct(ts.range[2]),"%Y-%m-%d %H:%M"),
                  "','YYYY-MM-DD HH24:MI')",sep=""))
  
  ## gestisce flag
  if(!is.null(flg) | flg.null) {
    CF <- NULL
    if(!is.null(flg)) {
      cf <- paste("(",
                  paste(paste("FLG_A=",flg,sep=""),collapse=" OR "),
                  ")",sep="")
      CF <- c(CF,cf)
    }        
    if(flg.null) {
      cf <- "(FLG_4 IS NULL)"
      CF <- c(CF,cf)
    }
    crit <- c(crit,paste("(",paste(CF, collapse=" OR "),")",sep=""))
  }
  if(!is.null(flg.excl)) {
    crit <- c(crit,
              paste("(",
                    paste(paste("FLG_A!=",flg.excl,sep=""),collapse=" AND "),
                    ")",sep=""))
  }        
  
  ## estrae dati
  query <- simple.query(tab,what,crit)
  if(verbose)  print(query)
  data <- dbGetQuery(con, query, ...)
  
  # arrotondamento in visualizzazione
  data$VALORE <- dbqa.round(x=data$VALORE, id.param=id.param)
  return(data)
}

## dato ID parametro, stazione, data inizio e fine,
## ottiene tutti gli ID sensore corrispondenti
dbqa.get.idcfgsens <- function(con,
                               id.param,
                               i.date=NULL,
                               f.date=NULL,
                               id.staz) {
  tab <- "AA_ARIA.ANG_CONFIG_SENSORI"
  crit <- c(paste("ID_PARAMETRO=",id.param,sep=""),
            paste("ID_STAZIONE=",id.staz,sep=""))
  if(!is.null(i.date)) {
    crit <- c(crit,
              paste("(DATA_FINE is null OR to_char(DATA_FINE,'YYYYMMDDHH')>='",
                    format(as.POSIXct(i.date),format="%Y%m%d%H"),
                    "')", sep=""))
  }
  if(!is.null(f.date)) {
    crit <- c(crit,
              paste("(DATA_INIZIO is null OR to_char(DATA_INIZIO,'YYYYMMDDHH')<='",
                    format(as.POSIXct(f.date),format="%Y%m%d%H"),
                    "')", sep=""))
  }
  #print(crit)
  query.idcfgsens <- simple.query(tab,what="ID_CONFIG_SENSORE",crit)
  #print(query.idcfgsens)
  idcfgsens <- dbGetQuery(con, query.idcfgsens)[,1]
  if(length(idcfgsens)>0) {
    query.idate <- simple.query(tab,what="DATA_INIZIO",crit)
    idate <- dbGetQuery(con, query.idate)[,1]
    idate <- pmax(as.POSIXct(idate),as.POSIXct(i.date))
    query.fdate <- simple.query(tab,what="DATA_FINE",crit)
    fdate <- dbGetQuery(con, query.fdate)[,1]
    fdate <- pmin(as.POSIXct(fdate),as.POSIXct(f.date))
    idx <- order(idate,decreasing=F)
    idate <- idate[idx]
    fdate <- fdate[idx]
    idcfgsens <- idcfgsens[idx]
  } else {
    idcfgsens=NULL
    idate=NULL
    fdate=NULL
  }
  
  out <- list(idcfgsens=idcfgsens, idate=idate, fdate=fdate)
  return(out)

}


## elenca le tabelle disponibili
dbqa.list.tables <- function(con) {
  tr <- dbSendQuery(con, "select table_name from all_tables")
  tables <- paste("AA_ARIA.",
                  sort(fetch(tr)[,1]),sep="")
  return(tables)
}

## elenca i campi disponibili in una tabella
dbqa.list.fields <- function(con,tab) {
  query <- paste("select * from ",tab," where rownum=0")
#  print(query)
  data <- dbGetQuery(con, query)
  out <- colnames(data)
  return(out)
}

## visualizza in tabella i parametri (inquinanti)
## con i relativi ID
dbqa.view.param <- function(con,FUN=View) {
  query <- "select ID_PARAMETRO,NOME from AA_ARIA.T$00$PARAMETRI"
  data <- fetch(ds <- dbSendQuery(con, query))
  idx <- order(data[,2])
  Data <- data[idx,]
  rownames(Data) <- rownames(data)
  FUN(Data)
}

dbqa.view.staz <- function(con,FUN=View) {
  query <- "select distinct ID_STAZIONE, NOME_STAZIONE, COMUNE, PROVINCIA, TIPOSTAZ, ZONA from AA_ARIA.ANG_CONFIG_SENSORI"
  data <- fetch(ds <- dbSendQuery(con, query))
  idx <- order(data[, 2])
  Data <- data[idx, ]
  rownames(Data) <- rownames(data)
  FUN(Data)
}

dbqa.isrrqa <- function(con,Id) {
  isr <- function(id) substr(dbGetQuery(con,paste("select aa_web.pk$99$zone.isRRQA_NEW (",
                                                  id,
                                                  ") from dual",
                                                  sep="")),
                             1,4)=="RRQA"
  mapply(FUN=isr, Id)
}

dbqa.list.active.staz <- function(con,
                                  prov=c("PC","PR","RE","MO","BO","FE","RA","FC","RN"),
                                  Day=Sys.Date(),
                                  mobile=FALSE) {
  day <- format(Day,format="%Y-%m-%d")
  query <- paste("SELECT ID_CONFIG_STAZ ",
                 "FROM aa_aria.VO$01$CONFIG_STAZIONI cst ",
                 "WHERE TRUNC (to_date('",
                 day,
                 "','YYYY-MM-DD')) BETWEEN cst.dth_i_vld ",
                 "AND NVL (cst.dth_f_vld, to_date('",
                 day,
                 "','YYYY-MM-DD')) ",
                 if(!mobile) "AND NVL (cst.flg_mobile, 0) = 0 ",
                 "AND COD_PRV IN (",
                 paste0("'",prov,"'",collapse=","),
                 ")",sep="")
  data <- dbGetQuery(con, query)
  Data <- as.character(sort(as.numeric(unlist(data))))
  names(Data) <- NULL
  return(Data)
}

## dato un ID parametro e stazione,
## scarica i dati disponibili dai sensori
## corrispondenti, così come sono
dbqa.get.datastaz <- function(con,
                              ts.range,
                              id.staz,
                              id.param,
                              flg=1,
                              tstep,
                              lod.manage=ifelse({id.param %in% c(12,14,15,18,29)},  #metalli e BaP
                                                "half",
                                                "keep"),
                              ...) {
  cfgsens <- dbqa.get.idcfgsens(con,
                                id.param,
                                i.date=ts.range[1],
                                f.date=ts.range[2],
                                id.staz)
  DATA <- NULL
  nsens <- length(cfgsens$idcfgsens)
  #print(lod.manage) 
  if(nsens>0) for (i in 1:nsens) {
    idate <- cfgsens$idate[i]
    fdate <- cfgsens$fdate[i]
    if(is.na(idate)) idate <- as.POSIXct(ts.range[1])
    if(is.na(fdate)) fdate <- as.POSIXct(ts.range[2])
    dat <- dbqa.get.datasens (con=con,
                              ts.range=c(idate,fdate),
                              id.cfgsens=cfgsens$idcfgsens[i],
                              id.param=id.param,
                              flg=flg,
                              ...)
    Dat <- dbqa.data2xts(dat)
    Dreg <- xts.regolarize(tstep,Dat,
                           f.time=as.POSIXct(ts.range[1], TZ="Africa/Algiers"),
                           l.time=as.POSIXct(ts.range[2], TZ="Africa/Algiers"))
    if(i>1) {
      DATA <- xts.blend(tstep, TZ="Africa/Algiers", DATA, Dreg)
    } else {
      DATA <- Dreg
    }
    #DATA <- data.frame(rbind(DATA,dat))
  }
  
  ## gestisce il LOD (metalli e BaP)
  if(lod.manage=="half") {
    lod <- dbqa.lod(con=con, id.param=id.param, days=index(DATA))
    DATA <- pmax(DATA,lod/2)
  }
  return(DATA)
}

## scrive o aggiorna valori nei campi di una tabella
dbqa.insert <- function(con, tab, values, columns=colnames(values), 
                        to_date=NULL,
                        update=F, verbose=F) {
  nv <- ncol(values)
  if(update) {
    action <- "update"
  } else {
    action <- "insert"    
  }
  vv <- paste(":",1:nv,sep="")
  vv[to_date] <- paste("TO_DATE(",vv[to_date],",'YYYY-MM-DD HH24:MI')",sep="")
  vvv <- paste(vv,collapse=", ")
  insStr <- paste(action," into ",
                  tab,
                  " (",
                  paste(columns,collapse=", "),
                  ") values (",
                  vvv,
                  ")",sep="")
  if(verbose)print(insStr)
  if(verbose)print(values)
  dbGetQuery(conn=con, statement=insStr, values)
}

## elimina record da una tabella
dbqa.delete <- function(con, tab, keys, values, verbose=F) {
  chkStr <- paste("select * from ",
                  tab,
                  " where ",
                  paste(paste(keys,values,sep="="),collapse=" AND "),
                  sep="")
  check <- dbGetQuery(conn=con, statement=chkStr)
  dbCommit(con)
  if(verbose) print(check)
  if(nrow(check)>0) {
    delStr <- paste("delete from ",
                    tab,
                    " where ",
                    paste(paste(keys,values,sep="="),collapse=" AND "),
                    sep="")
    if(verbose)print(delStr)
    dbGetQuery(conn=con, statement=delStr)    
  }
}

# arrotondamenti in visualizzazione, secondo indicazioni GdL
round_awayfromzero <- function(x,digits=0) trunc(x*10^digits+sign(x)*0.5)*10^-digits
dbqa.round <- function(x,id.param) {
  #                              O3 PM10 PM2.5 NO2  NOx SO2 C6H6   CO  BaP   As   Cd   Ni   Pb
  dat <- data.frame(idparam=c(   7,   5, 111,   8,   9,   1,  20,  10,  29,  18,  14,  15,  12),
                    digits= c(   0,   0,   0,   0,   0,   0,   1,   1,   4,   3,   3,   3,   6))
  idx <- which(id.param==dat$idparam)
  if(length(idx)==1) {
    dig <- dat$digits[idx]
    out <- round_awayfromzero(x,dig)    
  } else {
    out <- x
  }
  return(out)
}

# restituisce limite di quantificazione
dbqa.lod <- function(con, id.param, days=Sys.Date()) {
  days <- as.POSIXct(days)
  qqq <- paste0("select * from DETECTION_LIMIT where",
                " ID_PARAMETRO=",id.param)
  ddd <- dbGetQuery(con,qqq)
  matchRow <- function(day) {
    idx <- which(day >= ddd$DT_INIZIO_VALID & 
                   (is.na(ddd$DT_FINE_VALID) || day <= ddd$DT_FINE_VALID))
    if(length(idx)==0) idx <- NA
    return(idx[1])
  }
  idx <- unlist(lapply(days, matchRow))
  out <- ddd$DL[idx]
  return(out)
}


## converte nome inquinante in ID numerico
dbqa.get.idparam <- function(poll, con=NULL) {
  id.param <- NULL
  id.param <- switch(poll,
                     "SO2"=1,
                     "PM10"=5,
                     "PM2.5"=111,
                     "NO2"=8,
                     "NO"=38,
                     "NOx"=9,
                     "NOX"=9,
                     "CO"=10,
                     "O3"=7,
                     "BaP"=29,
                     "Ni"=15,
                     "As"=18,
                     "Cd"=14,
                     "Pb"=12)
  
  ## se non ? predefinito, lo cerca nei nomi del DB
  if(is.null(id.param)) {
    if(is.null(con)) {
      cat("Cannot search in the DB without 'con' argument",sep="\n")
    } else {
      ppp <- dbqa.view.param(con, FUN = return)
      fff <- agrep(poll, x = ppp$NOME, ignore.case = T)
      if(length(fff)==1) {
        cat(paste0("Selected '",ppp$NOME[fff],"'"),sep="\n")
        id.param <- ppp$ID_PARAMETRO[fff]
      } else if(length(fff)==0) {
        cat(paste("Cannot find",poll), sep="\n")
      } else if(length(fff)>1) {
        cat(paste0("Sorry, string '",poll,"' may refer to different IDs:"), sep="\n")
        cat(paste(paste0(ppp$ID_PARAMETRO,": '",ppp$NOME,"'")[fff],collapse="\n"),sep="\n")
        cat("",sep="\n")
      }
    }
  }
  if(is.null(id.param)) cat("Pollutant not managed",cat="\n")
  return(id.param)
}


## funzione di estrazione di un'elaborazione annuale
dbqa.get.elab <- function(con, 
                          year, 
                          id.param,
                          id.elab, 
                          type.elab="F",
                          only.rrqa=T, 
                          only.valid=T,
                          keep.all=F) {
  qqq <- paste("select * from WEB_STAT where ",
               "TO_CHAR(GIORNO,'YYYY')='",year,"' ",
               " and ID_PARAMETRO=",id.param,
               sep="")
  if(!is.null(id.elab)) qqq <- paste(qqq,"and ID_ELABORAZIONE=",id.elab)
  if(only.valid) qqq <- paste(qqq,"and FLG_ELAB=1")
  #print(qqq)
  dat <- dbGetQuery(con,qqq)
  
  # tiene solo RRQA se richiesto
  if(only.rrqa) {
    idx <- dbqa.isrrqa(con,dat$ID_CONFIG_STAZ)
    idx[is.na(idx)] <- FALSE
    if(length(idx)==0) return(NULL)
    dat <- dat[which(idx),]
  }
  # esclude doppioni piu' vecchi
  dat <- dat[order(dat$TS_UPD, decreasing=T),]
  dat <- dat[order(dat$TS_INS, decreasing=T),]
  idx <- paste(dat$ID_CONFIG_STAZ,dat$ID_ELABORAZIONE,dat$ID_EVENTO)
  dat <- dat[which(!duplicated(idx)),]
  
  
  if(keep.all) { # output originale completo
    out <- dat  
  } else {  # oppure colonne selezionate (pi? leggibile)
    out <- data.frame(ID_CONFIG_STAZ=dat$ID_CONFIG_STAZ, 
                      V_ELAB=dat[,paste("V_ELAB_",type.elab,sep="")],
                      ID_ELAB=dat$ID_ELABORAZIONE,
                      VERSION=dat$TS_INS,
                      FLAG_ELAB=dat$FLG_ELAB)
  }
  return(out)
}

## restituisce descrizione di una o pi? elaborazioni statistiche
dbqa.descr.elab <- function(con, id.elab=NULL) {
  qqq <- "select ID_ELABORAZIONE,DES_ELABORAZIONE from WEB_ELAB"
  dat <- dbGetQuery(conn = con,qqq)
  dat <- dat[match(id.elab,dat$ID_ELABORAZIONE), ,drop=F] ## riordina in base alla richiesta
  return(dat)
}


